Automattic/woocommerce-payments
# Adds Sofort payment gateway

Fixes #1515 

#### Changes proposed in this Pull Request


### Add Sofort Payment Gateway



This PR adds a working implementation for the [Sofort online banking payment gateway](https://stripe.com/docs/payments/sofort). Sofort is a redirection payment method, so a customer must be transported to their bank's payment portal to complete their transaction, at which point they can be redirected back to the respective store to confirm that the transaction has completed. It can take up to 14 days for a transaction to be confirmed, during which time the transaction will remain in a `processing` state. Stripe recommends that [a store should still fulfill any purchases during this `processing` state](https://stripe.com/docs/payments/sofort/accept-a-payment?platform=web#fulfillment), since on average only 0.2% of Sofort payment attempts will fail, so I suppose that's what we'll do.







#### Testing instructions





1. Please pull the changes from this PR and ensure you are using the latest version of the Payments Server.
2. Add the feature flag option `_wcpay_feature_sofort` to your site's `wp_options` table with a value of `1`.
3. Set your store's currency to use Euros.
4. Enable Sofort by visiting **Payments > Settings** and checking the checkbox at the bottom of the page titled `Enable/disable Sofort`.
5. Visit your store's shop as a customer, add an item to your cart, and proceed to checkout. When completing your billing details, you must ensure that the country is set to one of Austria (AT), Belgium (BE), Germany (DE), Spain (ES), Italy (IT), or the Netherlands (NL) as Sofort is only a valid payment method in these select countries.
6. For your payment method, choose Sofort, of course, and place your order.
7. You should eventually be redirected to a Stripe test page, where you have the option to authorize the payment or cause the transaction to fail. If you select to authorize the payment, you should be redirected to the completed checkout page. If you instead chose failure, you should be redirected back to the checkout screen with an error and the chance to try again.
8. Check the Order in `WP-Admin` and confirm that the correct notes about your transaction have been applied to your order.

-------------------
